### PR TITLE
Feature/update group buying order log

### PR DESCRIPTION
**Pull request by aniku777tw**

### CODE REVIEW

**Comment by aniku777tw** [comment]
[【團購】團購同行者可邀請非會員](https://trello.com/c/1mC3Fdj3/4372-%E3%80%90%E5%9C%98%E8%B3%BC%E3%80%91%E5%9C%98%E8%B3%BC%E5%90%8C%E8%A1%8C%E8%80%85%E5%8F%AF%E9%82%80%E8%AB%8B%E9%9D%9E%E6%9C%83%E5%93%A1)

File: src/order/order.controller.spec.ts [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+
+const appSetting = new AppSetting();
+appSetting.appId = app.id;
+appSetting.key = 'auth.service.client_id';
+appSetting.value = 'test';
+
+const appSecret = new AppSecret();
+appSecret.appId = app.id;
+appSecret.key = 'auth.service.key';
+appSecret.value = 'testKey';
```
Comment (jackey8616) [inline src/order/order.controller.spec.ts]: Use [this](https://github.com/urfit-tech/lodestar-server/blob/feature/update-group-buying-order-log/test/data.ts).

File: src/order/order.controller.spec.ts [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+  let manager: EntityManager;
+  let roleRepo: Repository<Role>;
+  let appPlanRepo: Repository<AppPlan>;
+  let appRepo: Repository<App>;
+  let appSecretRepo: Repository<AppSecret>;
+  let appSettingRepo: Repository<AppSetting>;
+  let memberRepo: Repository<Member>;
+  let orderLogRepo: Repository<OrderLog>;
+
+  beforeAll(async () => {
```
Comment (jackey8616) [inline src/order/order.controller.spec.ts]: Use `beforeEach` in order to clean datas in each cases.  
Then you can remove all 
  
```ts
await orderLogRepo.delete({});
await memberRepo.delete({});
```
  
in code.

File: src/order/order.controller.spec.ts [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    await roleRepo.save(role);
+    await appPlanRepo.save(appPlan);
+    await appRepo.save(app);
+    await appSecretRepo.save(appSecret);
+    await appSettingRepo.save(appSetting);
+
+    await application.init();
+  });
+
+  afterAll(async () => {
```
Comment (jackey8616) [inline src/order/order.controller.spec.ts]: `afterEach`.

File: src/order/order.controller.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  constructor(private orderService: OrderService) {}
+
+  @Put('transfer-received-order')
+  async transferOrder(@Body() dto: TransferReceivedOrderBodyDTO, @Headers() headers) {
+    const { token, memberId } = dto;
+    let transferOrderToken;
+
+    try {
+      const { authorization } = headers;
+      await verify(authorization.split(' ')[1], process.env.HASURA_JWT_SECRET);
```
Comment (jackey8616) [inline src/order/order.controller.ts]: process.env.HASURA_JWT_SECRET can use ConfigService to fetch.  
FYI: [AuthService](https://github.com/urfit-tech/lodestar-server/blob/feature/update-group-buying-order-log/src/auth/auth.service.ts#L18-L28).

File: src/order/order.controller.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    } catch {
+      throw new APIException({ code: 'E_TOKEN_INVALID', message: 'authToken is invalid' });
+    }
+
+    try {
+      transferOrderToken = (await verify(token, process.env.HASURA_JWT_SECRET)) as TransferReceivedOrderToken;
```
Comment (jackey8616) [inline src/order/order.controller.ts]: FYI: [AuthService](https://github.com/urfit-tech/lodestar-server/blob/feature/update-group-buying-order-log/src/auth/auth.service.ts#L18-L28).

File: src/order/order.controller.spec.ts [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    await orderLogRepo.delete({});
+    await memberRepo.delete({});
+    await appSettingRepo.delete({});
+    await appSecretRepo.delete({});
+    await appRepo.delete({});
+    await appPlanRepo.delete({});
+    await roleRepo.delete({});
+
+    application.close();
```
Comment (jackey8616) [inline src/order/order.controller.spec.ts]: await application.close();

File: src/order/order.controller.spec.ts [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+
+      await request(application.getHttpServer())
+        .put(apiPath.order.transferReceivedOrder)
+        .set(requestHeader)
+        .send(requestBody)
+        .expect(400)
+        .expect(/{"statusCode":400,"message":"E_TOKEN_INVALID"/);
+    });
+
+    it('should orderId is invalid', async () => {
```
Comment (jackey8616) [inline src/order/order.controller.spec.ts]: `Should raise error due to orderId is not found` ?

File: src/order/order.controller.spec.ts [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+        .put(apiPath.order.transferReceivedOrder)
+        .set(requestHeader)
+        .send(requestBody)
+        .expect(400)
+        .expect(/{"statusCode":400,"message":"E_NULL_ORDER"/);
+      await orderLogRepo.delete({});
+      await memberRepo.delete({});
+    });
+
+    it('should order received transfer success', async () => {
```
Comment (jackey8616) [inline src/order/order.controller.spec.ts]: `Should transfer received order successfully` ?

File: src/order/order.controller.spec.ts [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+      const orderData = {
+        appId: app.id,
+        email: '',
+        orderLogId: orderLog.id,
+        orderProductId: null,
+        title: 'test product',
+        ownerName: firstMember.name,
+      };
+
+      const orderDataToken = await sign(orderData, process.env.HASURA_JWT_SECRET, {
```
Comment (jackey8616) [inline src/order/order.controller.spec.ts]:
```ts
describe('OrderController (e2e)', () => {
  let configService: ConfigService<{HASURA_JWT_SECRET: string; }>;
  ...
  beforeEach(async () => {
    ...
    application = module.createNestApplication();
    configService = application.get<ConfigService<{ HASURA_JWT_SECRET: string}>>(ConfigService);
    await application.init();
    ...
  });

  ...

  it('should order received transfer success', async () => {
    ...
    const orderDataToken = await sign(orderData, configService.get('HASURA_JWT_SECRET'), {
        expiresIn: '30days',
    });
    ...
  });
});
```

File: src/order/order.controller.ts [hunk 10]
```diff
@@ -0,0 +1,38 @@
+import { Body, Controller, Headers, Put } from '@nestjs/common';
+import { verify } from 'jsonwebtoken';
+import { APIException } from '~/api.excetion';
+import { OrderService } from './order.service';
+import { TransferReceivedOrderBodyDTO, TransferReceivedOrderDTO, TransferReceivedOrderToken } from './order.type';
+
+@Controller({
+  path: 'order',
```
Comment (jackey8616) [inline src/order/order.controller.ts]: `orders` ?

**Review by jackey8616 — CHANGES_REQUESTED** [review]

File: src/order/order.controller.spec.ts (hunk 1 above)
Comment (aniku777tw) [reply src/order/order.controller.spec.ts]: updated

File: src/order/order.controller.spec.ts (hunk 2 above)
Comment (aniku777tw) [reply src/order/order.controller.spec.ts]: updated

**Comment by aniku777tw** [comment]
updated

File: src/order/order.controller.spec.ts [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+const apiPath = {
+  auth: {
+    token: '/auth/token',
+  },
+  order: {
+    transferReceivedOrder: '/orders/transfer-received-order',
+  },
+};
+
+describe('OrderController (e2e)', () => {
```
Comment (jackey8616) [inline src/order/order.controller.spec.ts:31]: Can organize like [this](https://github.com/urfit-tech/lodestar-server/blob/feature/update-group-buying-order-log/test/auth.e2e-spec.ts#L29).

File: src/order/order.controller.spec.ts [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+  beforeEach(async () => {
+    const module: TestingModule = await Test.createTestingModule({
+      providers: [OrderService],
+      controllers: [OrderController],
+      imports: [ApplicationModule],
+    }).compile();
+
+    application = module.createNestApplication();
+    configService = application.get<ConfigService<{ HASURA_JWT_SECRET: string }>>(ConfigService);
+    manager = application.get<EntityManager>('phdbEntityManager');
```
Comment (jackey8616) [inline src/order/order.controller.spec.ts]: `manager = application.get<EntityManager>('phdbEntityManager');`  

Cange to

`manager = application.get<EntityManager>(getEntityManagerToken('phdb'));`

**Review by jackey8616 — CHANGES_REQUESTED** [review]

**Review by jackey8616 — APPROVED** [review]
LGTM.

**Pull request merged**